src/stream/src/executor/dispatch/output_mapping.rs

```diff
@@ -0,0 +1,223 @@
+    TypeMapping {
+        indices: Vec<usize>,
+        /// `None` means no type change for this column.
+        types: Vec<Option<(DataType, DataType)>>,
+    },
+}
+
+impl DispatchOutputMapping {
+    pub(super) fn from_protobuf(proto: PbDispatchOutputMapping) -> Self {
+        let indices = proto.indices.into_iter().map(|i| i as usize).collect();
+
+        if proto.types.is_empty() {
+            Self::Simple(indices)
+        } else {
+            let types = (proto.types.into_iter())
+                .map(|t| {
+                    t.upstream.and_then(|u| {
+                        let d = t.downstream.unwrap();
+                        // Do an extra filter to avoid unnecessary mapping overhead.
+                        if u == d {
+                            None
+                        } else {
+                            Some((u.into(), d.into()))
+                        }
+                    })
+                })
+                .collect();
+            Self::TypeMapping { indices, types }
+        }
+    }
+
+    /// Apply the mapping to the chunk.
+    pub(super) fn apply(&self, chunk: StreamChunk) -> StreamChunk {
+        match self {
+            Self::Simple(indices) => {
+                // Only eliminate noop update when the number of columns is reduced.
+                if indices.len() < chunk.columns().len() {
+                    chunk.project(indices).eliminate_adjacent_noop_update()
+                } else {
+                    chunk.project(indices)
+                }
+            }
+
+            Self::TypeMapping { indices, types } => {
+                let (ops, columns, visibility) = chunk.into_inner();
+
+                let mut new_columns = Vec::with_capacity(indices.len());
+                for (i, t) in indices.iter().zip_eq_fast(types) {
+                    let mut column = columns[*i].clone();
+
+                    if let Some((from_type, into_type)) = t {
+                        let mut builder = into_type.create_array_builder(column.len());
+                        for (datum, vis) in column.iter().zip_eq_fast(visibility.iter()) {
+                            if !vis {
+                                builder.append_null();
+                            } else {
+                                let datum = type_mapping::do_map(datum, from_type, into_type);
+                                builder.append(datum);
+                            }
+                        }
+                        column = builder.finish().into();
+                    }
+
+                    new_columns.push(column);
+                }
+
+                // Always eliminate noop update since there's always type change, and updates to some struct
+                // fields may not be visible to the downstream.
+                StreamChunk::with_visibility(ops, new_columns, visibility)
+                    .eliminate_adjacent_noop_update()
+            }
+        }
+    }
```

Comment:
What's the difference between `DispatchOutputMapping` and `ProjectExecutor`? I guess we can reuse the `ProjectExecutor`? Speaking of this, is it also possible to implement ProjectExecutor in such two arms, one for `chunk.project`-only transformation, one for more complicated `NonStrictExpression` evaluation?

Reply:
You're right that we can definitely reuse `Project` here. However, given the logic is so specialized and straightforward, I don't think reusing it will simplify the implementation, not to mention that there will also be overhead for expr impl dispatching.